### PR TITLE
Add DisableSchemaIntrospection and ExecWithOptions

### PR DESCRIPTION
**Pull request by tatatodd**
This PR adds a new `DisableSchemaIntrospection` option.  The logic to handle this already existed, but was only controllable via the `DisableIntrospection` option, which also `__typename` introspection, which doesn't work for our purposes since we need it at runtime for union types.  That's basically a 2-line change in `selected.go` (the last file), with associated plumbing.

I've also added `ExecWithOptions`, which lets us delay choosing the option to query-execution time, rather than schema-parsing time.  We need this choice to be dynamic for our "debug users".  I did that the simplest way I could think of, with the caveats mentioned in the comments.

Finally, I realized I had a bug in my previous PR that added the validation cache wrt variables.  I.e. we need to verify variables even on cache hits, since they change per-query.  And we need to explicitly *not verify* in the `Schema.Validate` codepath, to ensure schemas that require non-null variables don't trivially fail validation.

All graphql-go continue to pass with this PR, as do the new tests I've added to test the above options.

### CODE REVIEW

File: internal/exec/selected/selected.go [hunk 1]
```diff
@@ -89,7 +90,7 @@ func applySelectionSet(r *Request, s *resolvable.Schema, e *resolvable.Object, s
 				}
 
 			case "__schema":
-				if !r.DisableIntrospection {
+				if !r.DisableIntrospection && !r.DisableSchemaIntrospection {
```
Comment (clischeske) [inline internal/exec/selected/selected.go]:
```
					if !r.DisableIntrospection || !r.DisableSchemaIntrospection {
```

Shouldn't this be `||` and not `&&`?

**Review by clischeske — APPROVED** [review]
This is super exciting. Considering merging it upstream?

File: internal/exec/selected/selected.go (hunk 1 above)
Comment (tatatodd) [reply internal/exec/selected/selected.go]: No, it should be `&&`, but it took me a while to work it out too.  It gave me motivation to get the tests plumbed through to make sure.  :)

I've rewritten it via deMorgans law to the possibly easier-to-read variant that uses `||`:
```
if !(r.DisableIntrospection || r.DisableSchemaInspection) {
```

Basically we only want to append the fields if both `Disable*` options are false.

**Review by tatatodd — COMMENTED** [review]
Sorry for the bait-and-switch, but I realized that my previous PR that added the validation cache was actually buggy; it wasn't handling variables correctly.

I've fixed it up in this PR, and we'll plow ahead.  Unfortunately that means validation will become slightly more expensive again, but I think (hope) it'll be fine.

WRT upstream merge, yup I promised Asim+Matt that I'd do that, but I never did, and we're kinda drifting further from upstream with our fork.  The good news is that there haven't been any upstream changes since November.  ;)

**Pull request merged**